Add Randomness how-to (#1484)
iotaledger/iota-wiki: docs/build/isc/v1.0.0-rc.6/docs/how-tos/get-randomness-on-l2.md

---

> description: The ISC Magic Contract allows EVM contracts to access ISC functionality.

Reviewer: ```suggestion
description: You can use the ISC Magic Contract in EVM contracts to access ISC functionality, such as randomness.
```

image: /img/logo/WASP_logo_dark.png
tags:

>   - configure
>   - using
>   - EVM
>   - magic
>   - Ethereum
>   - Solidity
>   - metamask
>   - JSON
>   - RPC

Reviewer: these seem out of place. how about:
```suggestion
  - magic contract
  - randomness
  - EVM
  - Ethereum
  - Solidity
  - ISC
```

---

# Get Randomness on L2

ISC introduces a feature not found on many other smart contract protocols: randomness.

> In the ISC protocol, for each state update each validator proposes a batch of smart contract requests they would like to process next. They commit to their proposals with a signature, after which a common subset of requests is found and a combined signature is produced.

Reviewer: ```suggestion
In the ISC protocol, for each state update, each validator proposes a batch of smart contract requests that they would like to process next. They commit to their proposals with a signature, after which a common subset of requests is found, and a combined signature is produced.
```

This combined signature is unpredictable, and not only serves as protection against MEV by pseudo-randomly ordering requests, but also provides a source of verifiable entropy for randomness on L2.

> This guide will show ways you can use this entropy to generate random values in your contracts.

Reviewer: ```suggestion
This guide will show you how you can use this entropy to generate random values in your contracts.
```

        bytes32 nonce = getNonce();
        bytes32 digest = keccak256(bytes.concat(entropy, nonce));

        uint256 value = uint256(digest);

        emit Int(value);
        return value;
    }

    function getBytes(uint length) public returns (bytes memory) {
        bytes32 entropy = ISC.sandbox.getEntropy();
        bytes32 nonce = getNonce();
        bytes32 digest = keccak256(bytes.concat(entropy, nonce));

        bytes memory value = new bytes(length);
        for (uint i = 0; i < length; i += 32) {
            digest = keccak256(bytes.concat(digest));
            for (uint j = 0; j < 32 && i + j < length; j++) {
                value[i + j] = digest[j];
            }
        }

        emit Bytes(value);
        return value;
    }
}

```

## Explanation

When you want to generate multiple random values within a single request, you need to take into account that entropy is constant within a request. In this contract we use an increasing nonce in addition to the entropy, to make sure we are generating unique values:

```solidity
uint256 private _nonce;

function getNonce() internal returns (bytes32) {
    return bytes32(_nonce++);
}
```

> ### Generating integers

Reviewer: ```suggestion
### Generating Integers
```
